_posts/2020-04-19-Julia-Antipatterns.md

```diff
@@ -0,0 +1,883 @@
+
+## Over-constraining argument types
+
+I will begin with a bold claim:
+Type constraints in Julia are **only** for dispatch.
+If you don't have multiple methods for a function, you don't need any type-constraints.
+If you must add type-constraints (for dispatch) do so as loosely as possible.
+I will justify this claim in the following sections.
+
+Note: while type constraints are only for dispatch, that does not mean the can't be used for other things.
+And in fact can be used for other things successfully even.
+But it doesn't change that the purpose of type-constraints is for dispatch: after all I can use a hammer as a paper-weight, but a hammer is still hammering nails.
+
+
+### Reasons people do this (and why they are wrong):
+I think the over-constraining of argument types comes from three main places.
+The belief it would make the code faster (false), safer (mostly false), or easier to understand (true, but...).
+The first two come from different languages which do not act like Julia.
+The last point on ease of understanding is absolutely true, but not worth it most of the time.
+There are better ways.
+
+The belief that adding type-constraints makes code faster, comes from a misunderstanding of Julia's compiler works.
+Julia's JIT compiler makes the code fast regardless of the type-constraints ([with a few limited exceptions](https://docs.julialang.org/en/latest/manual/performance-tips/#Be-aware-of-when-Julia-avoids-specializing-1)).
+Specialization is how some other languages use type-annotations  for performance, but Julia applies that technique all the time (and just in time).
+The standard action  of the Julia JIT is to specializes every function on the types of all arguments (not the type-constraints of the method, the types of the argument).
+This means it generates different machine code that is more optimally suited to the particular types.
+This includes things like removing branches that can't be met by this type, static dispatches, as well as actually better CPU instructions than a normal dynmaic language might use.
+One can see this change by comparing `@code_typed ((x)->2x)(1.0)` vs `@code_typed ((x)->2x)(0x1)`.
+Some languages, for example [Cython](https://cython.readthedocs.io/en/latest/src/quickstart/cythonize.html#faster-code-via-static-typing), *do* become much faster with type-annotations, as they do not have a JIT specializing every function on input type when it occurs.
+They do their code generation ahead of time so either have to handle all cases (if not specified) or can optimize for a particular (if specified).
+In Julia the code generated for a function will be just as fast with or without type-constraints.
+Another possible reason, not related to other languages, for this is misunderstanding [this part of the documentation](https://docs.julialang.org/en/v1/manual/performance-tips/), which applies to `struct` fields, not arguments.
```

Comment:
I think you wanted to link to this section? https://docs.julialang.org/en/v1/manual/performance-tips/#Type-declarations-1

Reply:
fixed, thanks